python/hopsworks_common/core/search_api.py

```diff
@@ -0,0 +1,304 @@
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+import logging
+from typing import Literal, get_args
+
+from hopsworks_common import client, search_results
+
+
+DOC_TYPE_ARG = Literal[
+    "FEATUREGROUP", "FEATUREVIEW", "TRAININGDATASET", "FEATURE", "ALL"
+]
+FILTER_BY_ARG = Literal["tag", "tag_name", "tag_key", "tag_value", "keyword"]
+
+
+class SearchApi:
+    def __init__(self):
+        self._log = logging.getLogger(__name__)
+
+    def featurestore_search(
+        self,
+        search_term: str,
+        filter_by: FILTER_BY_ARG = None,
+        offset: int = 0,
+        limit: int = 100,
+    ):
+        """
+        Search for feature groups, feature views, training datasets and features.
+        ```python
+
+        import hopsworks
+
+        project = hopsworks.login()
+
+        search_api = project.get_search_api()
+
+        result = search_api.featurestore_search("search-term")
+
+        # get feature group instance
+        featuregroup = result.featuregroups[0].get_feature_group()
+
+        ```
+        # Arguments
+            search_term: the term to search for.
+            filter_by: filter results by a specific field (default is None).
+            offset: the number of results to skip (default is 0).
+            limit: the number of search results to return (default is 100).
+        # Returns
+            `FeaturestoreSearchResult`: The results.
+        # Raises
+            `ValueError`: If the search term is not provided or if the filter_by is not one of the allowed values.
+            `hopsworks.client.exceptions.RestAPIError`: If the backend encounters an error when handling the request
+        """
+        return self._search(
+            search_term, "ALL", filter_by=filter_by, offset=offset, limit=limit
+        )
+
+    # FEATUREGROUP
+    def featuregroup_search(
+        self,
+        search_term: str,
+        filter_by: FILTER_BY_ARG = None,
+        offset: int = 0,
+        limit: int = 100,
+    ):
+        """
+        Search for feature group.
+        ```python
+
+        import hopsworks
+
+        project = hopsworks.login()
+
+        search_api = project.get_search_api()
+
+        result = search_api.featuregroup_search("search_term")
+
+        # get feature group instance
+        featuregroup = result.featuregroups[0].get_feature_group()
+
+        ```
+        # Arguments
+            search_term: the term to search for.
+            filter_by: filter results by a specific field (default is None).
```

Comment:
I think we should mention what are the possible filter values in the docstring.

Reply:
@ErmiasG are you going to add here, similar to the comment above?